packages/apollo-language-server/src/utilities/source.ts

```diff
@@ -181,3 +185,39 @@ export function getASTNodeAndTypeInfoAtPosition(
     return null;
   }
 }
+
+export function isFieldResolvedLocally(
+  fieldNode: FieldNode,
+  root: ASTNode
+): boolean | null {
+  const { loc } = fieldNode;
+  if (!loc) return null;
+
+  let underClientDirective: boolean = false;
+
+  function visitor(
```

Comment:
Some comments explaining the logic of this function would be very helpful. The way we're leveraging location may be confusing. Explaining the problem you ran into when solving this as well as the unintended perf benefit of not recursing are worth mentioning.